### PR TITLE
[DIR-197] new services page

### PR DESCRIPTION
**Pull request by stefan-kracht**
This PR implements the services page in the new UI

### [Preview 🚀](https://direktiv-ui-refs-pull-446-merge.direktiv.dev/stefan/explorer/tree)

I added a description of how it works and a lot of possible test cases in the [E2E Tests Ticket](https://linear.app/direktiv/issue/DIR-757/services-page).

**!! The API is not very reliable currently, some things might not work properly and could not finally tested in detail. I added a [ticket](https://linear.app/direktiv/issue/DIR-758/services-page-needs-more-testing-when-api-works-reliable) for that !!** 

### Notes

- The component for creating a service and a revision are very similar and could technically be merged into one. But at the moment both have a pretty different prop interface and since the mutation and query hooks can not be called conditionally these hooks must be hoisted, which currently adds a level of complexity. So a future abstraction could be to somehow only abstract the visual parts (the form) into a common component. 

### Other things out of services scope

- updated the `.vscode` folder rule in `.gitgnore` (ignore all, except `.vscode/extensions.json`) to allow other extensions to use that folder and not track it in this repo
- updated the `FormErrors` component to handle an empty `key` string (`src/pages/namespace/Services/Detail/Create.tsx` adds a general error message that is not linked to a certain field)
- Made popover in `ChildInstances` wider to support longer names
- Added an `ErrorPage` to all routes
- Added the possibility to use our Log component without timestamps
- `NoResults` component
  - Added Button support (and used it in empty directories)
  - Made Icon bigger and light gray
  - Added an Icon to the `NoResult` instance in the Variables-, Secrets- and Registries-Component
- updated `yarn run check` to run all checks in parallel (saves 10 seconds, and has the advantage of always seeing all check results and not the results until the point of the first one failing)
- adjusted one test after the new API was deployed (4f7dfd9792ca010e47d6a9b7ca2c3a3083fd46bd). The shape of a server result seems to have changed (for the better), no breaking change.
- Made Run Workflow button responsive <details><summary>show screenshots</summary> 
  Before
  CleanShot 2023-08-23 at 14 08 40
  After
  CleanShot 2023-08-23 at 14 08 14
</details> 

### Follow-Up Tickets
- [E2E Tests](https://linear.app/direktiv/issue/DIR-757/services-page)
- [Services page needs more testing when API works reliable](https://linear.app/direktiv/issue/DIR-758/services-page-needs-more-testing-when-api-works-reliable)

### CODE REVIEW

**Review by sebxian — CHANGES_REQUESTED** [review]
The services page and sub pages look good and seem to work as they should.

I must admit, however, that with my limited knowledge and limited time I can invest right now this is only a somewhat superficial impression. It would be very important to have an expert look at this as soon as possible and test it more thoroughly (after merging?).

I have made some suggestions in the code, mostly regarding spelling/readability.

File: src/api/services/__tests__/schema.test.ts [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+      expect(serviceNameSchema.safeParse("some/folder").success).toBe(false);
+      expect(serviceNameSchema.safeParse("some/").success).toBe(false);
+      expect(serviceNameSchema.safeParse("/folder").success).toBe(false);
+    });
+
+    test("can not contain uppercase characters middle", () => {
+      expect(serviceNameSchema.safeParse("fOldername").success).toBe(false);
+    });
+
+    test("can not end with various characters that are not lowercase characters or a digit", () => {
```
[the remark this answers is not in the document]
Comment (stefan-kracht) [reply src/api/services/__tests__/schema.test.ts]: I added 1 😊  6440dd7ae57d403b8cec9a6c337e450031503f31

File: src/api/services/query/getAll.ts [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+
+const fetchServices = async ({
+  queryKey: [{ apiKey, namespace }],
+}: QueryFunctionContext<ReturnType<(typeof serviceKeys)["servicesList"]>>) =>
+  getServices({
+    apiKey,
+    urlParams: { namespace },
+  }).then((res) => ({
+    // TODO: this should be changed in the backend
+    // reverse the order of functions (newer first)
```
[the remark this answers is not in the document]
Comment (stefan-kracht) [reply src/api/services/query/getAll.ts]: Yes, this is still an issue I wrote a [ticket](https://linear.app/direktiv/issue/DIR-784/change-order-of-services) and mentioned it in the comment (0f2341e49d4234567bc751c132e863b3ae7416f8)

File: src/pages/namespace/Services/Detail/Create.tsx [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+          if (defaultValues) {
+            return Object.keys(defaultValues).some((key) => {
+              const typedKey = key as keyof typeof defaultValues;
+              return x[typedKey] !== defaultValues[typedKey];
+            });
+          }
+          return true;
+        },
+        {
+          message: t("pages.services.revision.create.noChanges"),
```
[the remark this answers is not in the document]
Comment (stefan-kracht) [reply src/pages/namespace/Services/Detail/Create.tsx:77]: I like that idea, we should definitely introduce that

File: src/pages/namespace/Services/Detail/Create.tsx [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+        }
+        /**
+         * when no default values are available, it problably means that
+         * there is no previous revision. In this case, we don't let the
+         * user create one
```
[the remark this answers is not in the document]
Comment (stefan-kracht) [reply src/pages/namespace/Services/Detail/Create.tsx:82]: Yes, we could phrase it like that (1b3a96473fad988b4d52eb873f4c03f46fb108ee)

File: src/pages/namespace/Explorer/Tree/NoResult.tsx [hunk 5]
```diff
@@ -22,48 +23,52 @@ const NoResult: FC = () => {
...
-          >
-            <Button variant="outline">
-              <Folder />
-              {t("pages.explorer.tree.list.empty.createDirectory")}
-            </Button>
-          </DialogTrigger>
+    <div className="flex flex-col items-center gap-y-5">
+      <NoResultContainer
+        icon={FolderOpen}
+        button={
```
[the remark this answers is not in the document]
Comment (stefan-kracht) [reply src/pages/namespace/Explorer/Tree/NoResult.tsx]: Very good idea (3854b37f513d2969f745c448c09ca60770d458f3)!

File: src/componentsNext/FormErrors/index.tsx [hunk 6]
```diff
@@ -10,12 +10,12 @@ type FormErrorsProps = ComponentProps<typeof Alert> & { errors: errorsType };
...
   const { t } = useTranslation();
   const entries = Object.entries(errors);
-
   return entries.length ? (
     <Alert variant="error" {...props}>
       <ul>
         {entries.map(([key, value]) => (
-          <li key={key}>{`${key}: ${
+          // note: key might also be an empty string
+          <li key={key}>{`${key && `${key} :`} ${
```
[the remark this answers is not in the document]
Comment (stefan-kracht) [reply src/componentsNext/FormErrors/index.tsx]: Yes, that's a very good catch! (035e178ea2e8aff014e67bfd68b2e02d0ef772cc)

File: src/api/services/schema/revisions.ts [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+] as const;
+
+const RevisionConditionSchema = z
+  .object({
+    name: z.enum(revisionConditionNames),
+    status: StatusSchema,
+    reason: z.string(),
+    message: z.string(),
+  })
+  // handle status: False with reason: NoTraffic as status: True
```
[the remark this answers is not in the document]
Comment (stefan-kracht) [reply src/api/services/schema/revisions.ts]: Yes, this was not very well explained, I hope this is better: 566804dbb7f98f9a77660191c4104fefdc8182a9

File: src/api/services/query/getDetails.ts [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+     */
+    revisions: (oldData.revisions ?? [])
+      // swap the element that came in (if it already is in the cache)
+      .map((rev) => ({
+        /**
+         * we need to spread the old revision, because we don't consume all fields
+         * when streaming. The streaming payload has some minor inconsistencies
+         * with the revision schema. However, the fields that we keep from the
+         * cache are long living oney like the creation date and the name (which
+         * acts like an id)
```
[the remark this answers is not in the document]
Comment (stefan-kracht) [reply src/api/services/query/getDetails.ts]: fixed in d0fa1742611e6a9894d39d4128ccf3cc034ace02

**Comment by stefan-kracht** [comment]
Thanks for your review @sebxian. I somehow have to reduce the need of language cop in the future :D

- [one minor unresolved issue, that you could check](https://github.com/direktiv/direktiv-ui/pull/446/files/4f7dfd9792ca010e47d6a9b7ca2c3a3083fd46bd#r1307692524) 
- I also 
  - merged the base branch back in and fixed a merge conflict
  - updated an unrelated test according to your language suggestions (6afe1f7e6cd3768b79a58dfbe0d0970166c5c58e)

**Review by sebxian — APPROVED** [review]
Thanks for the updates, @stefan-kracht, they all look good to me!

**Pull request merged**